ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt

```diff
@@ -200,32 +241,88 @@ internal class PaywallViewModelImpl(
+    private suspend fun performPurchase(activity: Activity, packageToPurchase: Package) {
+        try {
+            val customPurchaseHandler = myAppPurchaseLogic?.let { it::performPurchase }
+
+            when (purchases.purchasesAreCompletedBy) {
+                PurchasesAreCompletedBy.MY_APP -> {
+                    checkNotNull(customPurchaseHandler) {
+                        "myAppPurchaseLogic must not be null when purchases.purchasesAreCompletedBy " +
+                            "is PurchasesAreCompletedBy.MY_APP"
+                    }
+                    when (val result = customPurchaseHandler.invoke(activity, packageToPurchase)) {
+                        is MyAppPurchaseResult.Success -> {
+                            purchases.syncPurchases()
+                            Logger.d("Dismissing paywall after purchase")
+                            options.dismissRequest()
+                        }
+                        is MyAppPurchaseResult.Cancellation -> {
+                            trackPaywallCancel()
+                        }
+                        is MyAppPurchaseResult.Error -> {
+                            result.errorDetails?.let { _actionError.value = it }
+                        }
+                    }
+                }
+                PurchasesAreCompletedBy.REVENUECAT -> {
+                    listener?.onPurchaseStarted(packageToPurchase)
+                    if (customPurchaseHandler != null) {
+                        Logger.e(
+                            "myAppPurchaseLogic expected to be null " +
+                                "when purchases.purchasesAreCompletedBy is .REVENUECAT. \n" +
+                                "myAppPurchaseLogic.performPurchase will not be executed.",
+                        )
+                    }
+                    val purchaseResult = purchases.awaitPurchase(
+                        PurchaseParams.Builder(activity, packageToPurchase),
+                    )
+                    listener?.onPurchaseCompleted(purchaseResult.customerInfo, purchaseResult.storeTransaction)
+                    Logger.d("Dismissing paywall after purchase")
+                    options.dismissRequest()
+                }
+                else -> {
+                    Logger.e("Unsupported purchase completion type: ${purchases.purchasesAreCompletedBy}")
+                }
+            }
+        } catch (e: PurchasesException) {
+            if (e.code == PurchasesErrorCode.PurchaseCancelledError) {
+                trackPaywallCancel()
+                listener?.onPurchaseCancelled()
+            } else {
+                listener?.onPurchaseError(e.error)
+                _actionError.value = e.error
+            }
+        }
+
+        finishAction()
+    }
+
+    private fun validateState() {
+        if (purchases.purchasesAreCompletedBy == PurchasesAreCompletedBy.MY_APP && options.myAppPurchaseLogic == null) {
+            error(
```

Comment:
This would crash... I'd avoid crashing and just display some error message in the paywall by setting the `_state` to the error state, for example. Not sure what behaviour we have in iOS though...

Reply:
On iOS we add the red error banner if it's run in debug, and crash in release, the thought being that it is very bad to release anything where you think people can make purchases, but can't. However maybe that's too harsh? I updated Android so it always shows a dialog (pictured below) in debug or release. 

![Untitled 39](https://github.com/user-attachments/assets/e356659d-46a6-493f-bf4b-99fdc37abb70)